File: benchmarks/php_comparison/html.php
<?php

declare(strict_types=1);

function escape_text(string $value): string
{
    return htmlspecialchars(
        $value,
        ENT_QUOTES | ENT_SUBSTITUTE | ENT_HTML5,
        'UTF-8'
    );
}

function render_card(int $id, string $title, bool $active): string
{
    ob_start();
    ?><article class="card" data-id="<?= $id ?>"><h2><?=
        escape_text($title)
    ?></h2><p>Customer #<?= $id ?> is <?=
        $active ? 'active' : 'paused'
    ?>.</p></article><?php
    $output = ob_get_clean();
    if (!is_string($output)) {
        throw new RuntimeException('failed to finish HTML buffer');
    }
    return $output;
}

echo render_card(0, 'A&B <Aster>', true), "\n";

$total = 0;
$active = true;
for ($index = 0; $index < 200_000; $index++) {
    $total += strlen(render_card(
        $index,
        'A&B <Aster>',
        $active
    ));
    $active = !$active;
}
echo $total, "\n";

File: benchmarks/go_comparison/html.go
package main

import (
	"fmt"
	"html"
	"strconv"
	"strings"
)

func renderCard(id int64, title string, active bool) string {
	state := "paused"
	if active {
		state = "active"
	}
	var output strings.Builder
	output.Grow(128)
	output.WriteString(`<article class="card" data-id="`)
	output.WriteString(strconv.FormatInt(id, 10))
	output.WriteString(`"><h2>`)
	output.WriteString(html.EscapeString(title))
	output.WriteString(`</h2><p>Customer #`)
	output.WriteString(strconv.FormatInt(id, 10))
	output.WriteString(" is ")
	output.WriteString(state)
	output.WriteString(".</p></article>")
	return output.String()
}

func main() {
	fmt.Println(renderCard(0, "A&B <Aster>", true))

	var total int
	var index int64
	active := true
	for index < 200_000 {
		total += len(renderCard(index, "A&B <Aster>", active))
		active = !active
		index++
	}
	fmt.Println(total)
}
